File: routers/recommendations.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Dict, Any

router = APIRouter(prefix="/recommend", tags=["Recommendations"])

class ScoreRequest(BaseModel):
    items: List[Dict[str, Any]]
    weights: Dict[str, float]

@router.post("/score")
async def calculate_scores(data: ScoreRequest):
    """
    Calculate scores for a list of items based on provided weights.
    """
    if not data.items:
        return {"scored_items": []}
    
    scored_items = []
    for item in data.items:
        score = 0.0
        for key, weight in data.weights.items():
            val = item.get(key, 0)
            if isinstance(val, (int, float)):
                score += float(val) * weight
            elif isinstance(val, bool):
                score += (1.0 if val else 0.0) * weight
        
        scored_item = item.copy()
        scored_item["_score"] = round(score, 4)
        scored_items.append(scored_item)
    
    # Sort by score descending
    scored_items.sort(key=lambda x: x["_score"], reverse=True)
    
    return {"scored_items": scored_items}

File: src/routers/recommendations.ts
import express, { Router } from 'express';
import { ScoreRequestSchema } from '../types/schemas';
import { webhookLogger } from '../utils/logger';

const router: Router = express.Router();

/**
 * Calculate scores for a list of items based on provided weights.
 */
router.post('/score', (req, res) => {
  try {
    const data = ScoreRequestSchema.parse(req.body);

    if (!data.items || data.items.length === 0) {
      return res.json({ scored_items: [] });
    }

    const scoredItems = data.items.map((item) => {
      let score = 0.0;
      for (const [key, weight] of Object.entries(data.weights)) {
        const val = item[key];
        if (typeof val === 'number') {
          score += val * weight;
        } else if (typeof val === 'boolean') {
          score += (val ? 1.0 : 0.0) * weight;
        }
      }

      return {
        ...item,
        _score: Number(score.toFixed(4)),
      };
    });

    // Sort by score descending
    scoredItems.sort((a, b) => b._score - a._score);

    webhookLogger.log(`Calculated scores for ${scoredItems.length} items`);
    res.json({ scored_items: scoredItems });
  } catch (error: any) {
    webhookLogger.log(`Error calculating scores: ${error.message}`, 'ERROR');
    res.status(400).json({ detail: error.errors || error.message });
  }
});

export default router;
